### ISSUE TITLE
Interoperability of Find mode and Immediate mode API calls is limited in MP/MT environments.

### ISSUE THREAD

**Issue by atamazov** [ROCm/MIOpen#219]
The documentation does not allow cooperative use of Find and Immediate API calls so far. The results of Immediate mode API calls should not be passed to Find mode API calls and vice versa.

Find and Immediate modes are not fully interoperable due to API restrictions. For example, `Find()` returns an `algorithm`, but Immediate mode needs `solution_id` to run etc. 

However we would like to ensure interoperability (through Find 2.0 API, see #374). And in certain cases we can use Immediate mode calls within Find mode chains and vice versa even today (NB: unspecified behavior!)

The following limitation apply regardless of the above:
- In MP/MT environments, the results of Immediate mode API calls should not be passed into Find mode API calls.

Most likely, the limitation will be completely removed after switching to Find API 2.0.

----
[Informative] It is Ok to mix Immediate mode calls with Find calls provided that the results of one API are not passed to another API. For example, the following Immediate mode calls are Ok to mix with Find calls:
```
miopenConvolutionForwardGetSolutionWorkspaceSize()
miopenConvolutionBackwardDataGetSolutionWorkspaceSize()
miopenConvolutionBackwardWeightsGetSolutionWorkspaceSize()
```

`GetForwardSolutionWorkspaceSize()` works independently of find-db. So the use case scenario like this wouldn't fail:
```
... GetSolutions() ---> Find() --> GetSolutionWorkspaceSize(solution_id) ...
                       updates
                       find-db                    
```
As GetSolutionWorkspaceSize() does not read WS size from find-db but always _computes_ it, it returns valid value workspace size regardless of any find-db updates that may happen after GetSolutions().

**Comment by atamazov** [comment]
These calls are optional. The obvious proposal is to remove them.

**Comment by atamazov** [comment]
@pfultz2 @daniellowell @DrizztDoUrden FYI

**Comment by JehandadKhan** [comment]
> These calls are optional. The obvious proposal is to remove them.

I am not sure I follow, why is removal the obvious solution ? 

A simpler solution is to not store the workspace size in FDB ( I think we went over this earlier). Since the workspace size can be calculated by the solver without the perf params ( ie. a PDB access) GWSS calls should avoid the fdb access and just compute the workspace size from the problem and tensor descriptors.

**Comment by atamazov** [comment]
GWSS must know solver id to do this calculation.

**Comment by JehandadKhan** [comment]
@atamazov  Which should be known in immediate mode, even if its not, it can be enumerated from the any-solver containers.

**Comment by atamazov** [comment]
You are right. And I was a bit wrong in the description (partially fixed). I need to think about this...

**Comment by atamazov** [comment]
[Informative] `GetForwardSolutionWorkspaceSize()` works independently of find-db. Actually it MUST work this way due to some find-db limitations imposed by Find API 1.0. So the use case scenario like this would never fail:

## Use case
```
process 1 --> GetSolutions -------x------> GetSolutionWorkspaceSize(solution_id)
                                  ^
                   find-db update |
                                  |
process 2 --------------------> Find ---> ...
```
As GetSolutionWorkspaceSize does not read WS size from find-db but always _computes_ it, it returns valid value workspace size regardless of any find-db updates that may happen in parallel.

**Comment by atamazov** [comment]
@JehandadKhan I will update issue title & description, and hide or delete some irrelevant comments soon.

**Comment by JonathanLichtnerAMD** [comment]
Closing since MIOpen is being migrated to [ROCm/rocm-libraries](https://github.com/ROCm/rocm-libraries).  Please re-open there if you believe this is still an issue.
